Demo for intra process communication

wjwwood opened the pull request:
Work in progress, not ready for review.

Connects to ros2/ros2#59

dirk-thomas commented on intra_process_comms/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+ament_package()
+
+# This file provides the add_executable_for_each_rmw_impl function.
+include(cmake/add_executable_for_each_rmw_impl.cmake)
```
Please use ros2/rmw_implementation#3 instead.

dirk-thomas commented on intra_process_comms/package.xml:
```diff
@@ -0,0 +1,23 @@
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+
+  <buildtool_depend>rosidl_default_generators</buildtool_depend>
```
The package does not contain any messages and therefore neither requires `rosidl_default_generators` nor `rosidl_default_runtime`.

dirk-thomas commented on intra_process_comms/src/intrapc.cpp:
```diff
@@ -0,0 +1,72 @@
+
+void on_message(const std_msgs::msg::String::ConstSharedPtr & msg)
+{
+  std::cout << "I heard [" << msg->data << "]" << std::endl;
```
Please use `printf`.

Same below.

dirk-thomas commented on intra_process_comms/src/intrapc.cpp:
```diff
@@ -0,0 +1,72 @@
+  std::cout << "I heard [" << msg->data << "]" << std::endl;
+  rclcpp::sleep_for(1_s);
+  if (msg->data.at(msg->data.length() - 1) == '9')
+  {
```
Style.

dirk-thomas commented on intra_process_comms/src/intrapc.cpp:
```diff
@@ -0,0 +1,72 @@
+  rclcpp::init(argc, argv);
+
+  auto node1 = rclcpp::Node::make_shared("node1");
+  auto node2 = rclcpp::Node::make_shared("node2");
```
Don't they have to enable intra process comm here (like https://github.com/ros2/system_tests/pull/28/files#diff-fe711788916f61638c377d72e77584a5R28)?

dirk-thomas commented on intra_process_comms/src/intrapc.cpp:
```diff
@@ -0,0 +1,72 @@
+
+#include <std_msgs/msg/string.hpp>
+
+void on_message(const std_msgs::msg::String::ConstSharedPtr & msg)
```
The reference `&` is not needed for shared pointers.

Same below.

jacquelinekay replied:
Do you plan on adding output that shows intra-process communication is being used (like comparing the address of the message pointer)?

wjwwood replied:
> The reference & is not needed for shared pointers.

This is the signature for subscription callbacks, e.g.: https://github.com/ros2/examples/blob/master/rclcpp_examples/src/topics/listener.cpp#L22

Why not take them by reference?

dirk-thomas replied:
A reference is basically a pointer. If the caller lets the shared pointer go out of scope the reference this function is holding becomes invalid. Therefore I think the function should receive its own shared pointer which will then also contribute to the reference counting.

wjwwood replied:
Part of the contract of the subscription callback is that this will not happen. I'm not going to change this here until we change it in all of our callbacks.

wjwwood replied:
I changed the one I control here: cb42509

dirk-thomas replied:
Sure, I have created ros2/ros2#94 to keep track of fixing it in the existing code.

jacquelinekay commented:
The description says it's not ready for review, but I assume this is outdated because you've moved this set of changes into "In Review"...

wjwwood commented:
Yeah, this one is still pretty basic. I think it's worth reviewing now and merging and it will get expanded and updated when I write the demo tutorial.

I addressed the comments in the latest commits.

esteve commented:
+1
